sdk/storage/azure_storage_blob/src/partitioned_transfer/download.rs

```diff
@@ -0,0 +1,311 @@
+use std::{collections::VecDeque, ops::Range, task::Poll};
+
+use azure_core::http::AsyncRawResponse;
+use bytes::Bytes;
+
+use crate::models::content_range::ContentRange;
+
+use super::*;
+
+#[async_trait::async_trait]
+pub(crate) trait PartitionedDownloadBehavior {
+    async fn transfer_range(&self, range: Range<u64>) -> AzureResult<AsyncRawResponse>;
+}
+
+/// Holds either a future to be polled or its output to be persisted.
+/// Allows tracking pending futures and their completed results in the
+/// same data structure.
+enum PollPersist<Out, Fut> {
+    Ready(Out),
+    Pending(Fut),
+}
+type OpsFuture = Pin<Box<dyn Future<Output = Result<Bytes, azure_core::Error>>>>;
+
+/// Returns a stream that runs up to parallel-many ranged downloads at a time.
+///
+/// Downloads are stored in-order. The returned stream will produce an item only when the next
+/// download in the sequence has been buffered, regardless of the state of any other downloads.
+/// This means completed ranged downloads may sit for a while while earlier ones complete.
+///
+/// A download that has completed buffering but has not yet returned its buffer in the resulting
+/// stream will still count when determining current running operations. This is so the stream can
+/// promise its buffered bytes do not exceed parallel * partition_size.
+pub(crate) async fn download<'a, T: PartitionedDownloadBehavior>(
```

Comment:
Do we plan to offer a `DownloadTo` like API where we can take in a stream and seek+write the results to it. This feels like it would be more performant as we don't have to return bytes in order.